rfcs/0009-separate-wal.md

```diff
@@ -0,0 +1,99 @@
+# Separate Object Store for WAL
+
+Status: Draft
+
+Authors:
+
+* [Sergei Sitnikov](https://github.com/taburet)
+
+## Motivation
+
+Sequential writes are a common pattern. Combined with `await_durable` set to
+`true`, which is the default, WAL performance is the main limiting factor for
+the sequential writes throughput. Even if `flush_interval` is set to an extremely
+low value, it can't overcome the inherent latency of the underlying object store.
+
+The obvious solution is switching to a lower latency object store; for instance,
+from S3 to S3 Express One Zone. But lower latency stores usually come with a cost,
+they might be significantly more expensive or might be limited in features. Using
+such stores for the main store containing the entire database might be undesirable.
+
+## Goals
+
+- Support an optional separate object store dedicated specifically for WAL.
+- Provide a user-facing configuration mechanism.
+- Avoid significant refactorings.
+
+## Public API
+
+The only change on the public API level is configuration:
+
+```rust
+pub struct DbOptions {
+    // ...
+
+    /// An optional [object store](ObjectStore) dedicated specifically for WAL.
+    ///
+    /// If not set, the main object store passed to `Db::open(...)` will be used
+    /// for WAL storage.
+    ///
+    /// NOTE: WAL durability and availability properties depend on the properties
+    ///       of the underlying object store. Make sure the configured object
+    ///       store is durable and available enough for your use case.
+    #[serde(skip)]
+    pub wal_object_store: Option<Arc<dyn ObjectStore>>
```

Comment:
API for opening the dB `open_with_opts` takes an object store, and DbOptions.
WAL object store is passed inside DbOptions in this proposal.

It doesn't look uniform. Should we pass both the same way (either as parameter or as DbOption)?

Reply:
my reasoning was really unsophisticated, if it's optional it goes to options, but I'm fine with both approaches

Reply:
Let's make it passed into open_with_opts like the regular object store. That way it stays uniform. This is a breaking API change, which is fine (doesn't affect storage format).

Reply:
moved it to `open_with_opts` https://github.com/slatedb/slatedb/pull/533/commits/f10429b0e0546d6850ec7a3980b38e4e607750bd

Reply:
this seems to be resolved, but just another point is that DbOptions need to be serializable (which based on the whole other discussion in this RFC is another reason why it shouldn't be part of the options)

Reply:
Ya, I think we're gonna go with @calavera's RFC 10 (a builder). Should solve this. :)